### PR TITLE
Unify gRPC Gateways into Single, Shared lib

### LINKED ISSUE #8783: Unify gRPC Gateways Into Single, Shared Package
# 💎 Issue

### Background

We have several instaces of gRPC gateways which are essentially duplicated code, all throughout Prysm. We have one in the beacon node, one in the validator, and we should definitely create a shared/gateway package that contains this code in a single place to make it more maintanable.

### PR DESCRIPTION
**Pull request by ahadda5**
**What type of PR is this?**

Cleanup 

**What does this PR do? Why is it needed?**
Unify the gateway package which has  duplicated code funcs between the beacon-chain and validator into a shared gateway lib. 

**Changes**

- The shared lib implements a shared ServiceRegistry and overwrites its Start(), Stop() and status(). 
- Extract the beacon-chain gateway and validator gateway libraries and create the shared/gateway lib
- Moved the server folder under beacon-chain/gateway into its own folder. (main func is there)
- Add a caller id to the Gateway struct to distinguish between the two handlers(beacon-chain vs validator) on the Start() of the overloaded func (the code could be made cleaner ) 
-  Add remoteCerrt, enableDebugRPCEndpoints and maxCallRecvMsgSize to a combined Gateway struct, which are used by beacon-chain. 
- Combine the cors and handler files of beacon-chain into the shared/gateway
- Stop() and Status() were identical for beacon-chain and validator and merged 
- New() is divided into New beacon vs New Validator. Different signatures  
- Start() requires the different handlers registration of  beacon-chain vs validator. 
- validator/web  lib is imported to shared/gateway for the use in the validator lines of codes of  gateway. 

**Testing**
Both lib were ripped out of beacon-chain/gateway and validator/rpc/gateway respectively and placed under shared/gateway . Testing was done as follow 

1. bazel build  //shared/gateway:go_default_library 
2. bazel test //shared/gateway:go_default_test  --test_output=streamed
3. bazel build //beacon-chain/server:go_default_library  (main func)
4. bazel test //beacon-chain/node:go_default_test  --test_output=streamed
5. bazel run //beacon-chain/server (ensure gRPC is called) 
6. bazel test //validator/rpc:go_default_test  --test_output=streamed  (to ensure it still runs okay).
7. bazel test //validator/node:go_default_test  --test_output=streamed

**Which issues(s) does this PR fix?**

Fixes #8783

**Other notes for review**

- [x] Need to add **unit** test for this new shared lib.  
- [x] grpc.WithDialer is deprecated. Need to update

### CODE REVIEW

**Comment by ahadda5** [comment]
gazelle keeps changing my deps.  The one it changes to, fails build because the target does not exist. 
How can i prevent the switch 

```

(need this)        "@com_github_prysmaticlabs_ethereumapis//eth/v1alpha1:go_grpc_gateway_library",

(not this)      "@com_github_prysmaticlabs_ethereumapis//eth/v1alpha1_gateway:go_default_library",

```

**Comment by rkapka** [comment]
I am not sure why this happens, but you can try to manually edit the BUILD file and not run gazelle.

**Comment by rkapka** [comment]
What do you mean by "target does not exist"? The import is:
```
ethpb "github.com/prysmaticlabs/ethereumapis/eth/v1alpha1_gateway"
```
It looks to me as if gazelle is doing the right thing.

File: shared/gateway/gateway.go [hunk 1]
```diff
@@ -1,29 +1,33 @@
-// Package gateway defines a gRPC gateway to serve HTTP-JSON
-// traffic as a proxy and forward it to a beacon node's gRPC service.
```
Comment (rkapka) [inline shared/gateway/gateway.go:2]: Why did you remove the documentation? If something is out of date, please update it.

File: shared/gateway/gateway.go [hunk 2]
```diff
@@ -1,29 +1,33 @@
...
 )
 
 var _ shared.Service = (*Gateway)(nil)
 
-// Gateway is the gRPC gateway to serve HTTP JSON traffic as a proxy and forward
-// it to the beacon-chain gRPC server.
+// Gateway is the gRPC gateway to serve HTTP JSON traffic as a
+// proxy and forward it to the gRPC server.
 type Gateway struct {
+	callerId                string
```
Comment (rkapka) [inline shared/gateway/gateway.go:30]: Let's use an enumeration here. Example: https://github.com/prysmaticlabs/prysm/blob/648e3608429691e15ccbc0d62dc8089c4440c5e7/slasher/detection/attestations/types/types.go#L15

File: shared/gateway/gateway.go [hunk 3]
```diff
@@ -38,53 +42,131 @@ type Gateway struct {
 	maxCallRecvMsgSize      uint64
 }
 
-// Start the gateway service. This serves the HTTP JSON traffic on the specified
-// port.
+// New returns a new gateway server which translates HTTP into gRPC.
+// Accepts a context and optional http.ServeMux.
+func NewValidator(
+	ctx context.Context,
+	callerId string,
```
Comment (rkapka) [inline shared/gateway/gateway.go:49]: The `callerId` parameter doesn't make sense because `NewValidator` will always be called with `validator-rpc`. You can use the literal string value in the `&Gateway` expression.

File: shared/gateway/gateway.go [hunk 4]
```diff
@@ -38,53 +42,131 @@ type Gateway struct {
...
+		mux:            http.NewServeMux(),
+		allowedOrigins: allowedOrigins,
+	}
+}
+
+// New returns a new gateway server which translates HTTP into gRPC.
+// Accepts a context and optional http.ServeMux.
+func NewBeacon(
+	ctx context.Context,
+	callerId string,
```
Comment (rkapka) [inline shared/gateway/gateway.go:68]: Same comment about `callerId` as for `NewValidator`.

File: shared/gateway/gateway.go [hunk 5]
```diff
@@ -38,53 +42,131 @@ type Gateway struct {
 	maxCallRecvMsgSize      uint64
 }
 
-// Start the gateway service. This serves the HTTP JSON traffic on the specified
-// port.
+// New returns a new gateway server which translates HTTP into gRPC.
+// Accepts a context and optional http.ServeMux.
```
Comment (rkapka) [inline shared/gateway/gateway.go:46]: This function does not accept an `http.ServeMux`.

**Review by rkapka — CHANGES_REQUESTED** [review]

File: shared/gateway/gateway.go (hunk 1 above)
Comment (ahadda5) [reply shared/gateway/gateway.go:2]: Will restore.

File: shared/gateway/gateway.go (hunk 3 above)
Comment (ahadda5) [reply shared/gateway/gateway.go:49]: The challenge is not the New function, but the overridden SharedRegistry  Start, Status, and Stop funcs. While the latter two can be merged without harming the original beacon and Validator nodes, the former (Start) has diff code forks .

**Review by ahadda5 — COMMENTED** [review]
> What do you mean by "target does not exist"? The import is:
> 
> ```
> ethpb "github.com/prysmaticlabs/ethereumapis/eth/v1alpha1_gateway"
> ```
> 
> It looks to me as if gazelle is doing the right thing.

Gazelle is defaulting to the go_default_library.  That does not exist in that lib for some reason. 
I added gazelle ignore so not to change the build file. (same as the replaced beacon gateway build)

File: shared/gateway/gateway.go (hunk 2 above)
Comment (ahadda5) [reply shared/gateway/gateway.go]: done

**Comment by ahadda5** [comment]
Also getting a deprecated func in grpc. Will removing it affect the master branch? 

[grpc.WithDialer](https://github.com/prysmaticlabs/prysm/blob/d150f4714dfdda545d2b99b467224e4544d7c088/shared/gateway/gateway.go#L297)

**Comment by ahadda5** [comment]
I noticed that the grpc.WithDialer is deprecated. Need to update to . WithContextDialer . [Line 297](https://github.com/ahadda5/prysm/blob/d150f4714dfdda545d2b99b467224e4544d7c088/shared/gateway/gateway.go#L297)

Second, what is the effect of having a Gateway struct with no remoteCert, enableDebugRPCEndpoints, maxCallRcvmsgSize defined (nil) for the validator connections. Those are beacon's fields.   

The aim throughout this PR was to combine code.

**Comment by ahadda5** [comment]
Made cmd/beacon-chain/flags  visible to         "//shared/gateway:__pkg__".  Not sure if it is worth it considering how basic the test is.

File: cmd/beacon-chain/flags/BUILD.bazel [hunk 6]
```diff
@@ -12,6 +12,7 @@ go_library(
     visibility = [
         "//beacon-chain:__subpackages__",
         "//cmd/beacon-chain:__subpackages__",
+        "//shared/gateway:__pkg__",
```
Comment (ahadda5) [inline cmd/beacon-chain/flags/BUILD.bazel:15]: Needed to expose this for testing purposes, the flags for the context needed for a gateway.

File: shared/gateway/gateway_test.go [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+		enableDebugRPCEndpoints,
+		ctx.Uint64("grpc-max-msg-size"),
+	)
+
+	beaconGateway.Start()
+	go func() {
+		require.LogsContain(t, hook, "Starting gRPC gateway")
+	}()
+
+	err := beaconGateway.Stop()
```
Comment (rkapka) [inline shared/gateway/gateway_test.go:49]: Is there a way to verify that it actually stopped?

File: shared/gateway/gateway_test.go [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+		gatewayAddress,
+		allowedOrigins,
+	)
+
+	validatorGateway.Start()
+	go func() {
+		require.LogsContain(t, hook, "Starting gRPC gateway")
+	}()
+
+	err := validatorGateway.Stop()
```
Comment (rkapka) [inline shared/gateway/gateway_test.go:83]: Is there a way to verify that it actually stopped?

File: shared/gateway/gateway.go [hunk 9]
```diff
@@ -1,30 +1,45 @@
...
 	"google.golang.org/grpc/credentials"
 )
 
 var _ shared.Service = (*Gateway)(nil)
 
-// Gateway is the gRPC gateway to serve HTTP JSON traffic as a proxy and forward
-// it to the beacon-chain gRPC server.
+// CallerId defines whether the caller node is a beacon
+// or a validator node. This helps register the handlers accordingly.
+type CallerId string
```
Comment (rkapka) [inline shared/gateway/gateway.go:31]: It is more idiomatic to use an `int` enum just like in https://github.com/prysmaticlabs/prysm/blob/648e3608429691e15ccbc0d62dc8089c4440c5e7/slasher/detection/attestations/types/types.go#L15

File: shared/gateway/gateway.go [hunk 10]
```diff
@@ -1,30 +1,45 @@
...
+
+const (
+	ValidatorCaller = "validator-rpc"
+	BeaconCaller    = "beacon-rpc"
```
Comment (rkapka) [inline shared/gateway/gateway.go:35]: No need for the `Caller` suffix, it's redundant when the type name is `CallerId`.

File: shared/gateway/gateway.go [hunk 11]
```diff
@@ -38,53 +53,133 @@ type Gateway struct {
 	maxCallRecvMsgSize      uint64
 }
 
-// Start the gateway service. This serves the HTTP JSON traffic on the specified
-// port.
+// New returns a new gateway server which translates HTTP into gRPC.
+// Accepts a context.
+func NewValidator(
+	ctx context.Context,
+	callerId CallerId,
```
Comment (rkapka) [inline shared/gateway/gateway.go:60]: I still don't understand why you pass the `callerId` parameter here when it's always going to be `"validator-rpc"`. How about this:
```
func NewValidator(
	ctx context.Context,
	remoteAddress,
	gatewayAddress string,
	allowedOrigins []string,
) *Gateway {
	return &Gateway{
		callerId:       "validator-rpc",
		remoteAddr:     remoteAddress,
		gatewayAddr:    gatewayAddress,
		ctx:            ctx,
		mux:            http.NewServeMux(),
		allowedOrigins: allowedOrigins,
	}
}
``` 

Similarly for the beacon node.

File: shared/gateway/gateway_test.go [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+	"github.com/urfave/cli/v2"
+)
+
+// Test that beacon gateway Start, Stop.
+func TestBeaconGateway_StartStop(t *testing.T) {
+	hook := logTest.NewGlobal()
+
+	app := cli.App{}
+	set := flag.NewFlagSet("test", 0)
+	ctx := cli.NewContext(&app, set, nil)
```
Comment (rkapka) [inline shared/gateway/gateway_test.go:21]: I can't see why you need all these variables. Why can't you pass values directly into `NewBeacon`?

File: shared/gateway/gateway_test.go [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+
+}
+
+// Test that validator gateway Start, Stop.
+func TestValidatorGateway_StartStop(t *testing.T) {
+	hook := logTest.NewGlobal()
+
+	app := cli.App{}
+	set := flag.NewFlagSet("test", 0)
+	ctx := cli.NewContext(&app, set, nil)
```
Comment (rkapka) [inline shared/gateway/gateway_test.go:60]: Same comment as for the beacon test.

**Review by rkapka — CHANGES_REQUESTED** [review]

File: shared/gateway/gateway.go (hunk 9 above)
Comment (ahadda5) [reply shared/gateway/gateway.go:31]: Okay will adjust.

File: shared/gateway/gateway.go (hunk 10 above)
Comment (ahadda5) [reply shared/gateway/gateway.go:35]: Agreed!

File: shared/gateway/gateway.go (hunk 11 above)
Comment (ahadda5) [reply shared/gateway/gateway.go:60]: Agreed. redundant.

File: shared/gateway/gateway_test.go (hunk 7 above)
Comment (ahadda5) [reply shared/gateway/gateway_test.go:49]: Unless we add a log like the Start does?

File: shared/gateway/gateway_test.go (hunk 12 above)
Comment (ahadda5) [reply shared/gateway/gateway_test.go:21]:
```
beaconGateway := NewBeacon(
		ctx.Context,
		BeaconCaller,
		selfAddress,
		selfCert,
		gatewayAddress,
		nil, /*optional mux*/
		allowedOrigins,
		enableDebugRPCEndpoints,
		ctx.Uint64("grpc-max-msg-size"),
	)
```

Will be 

```
beaconGateway := NewBeacon(
		cli.NewContext(&cli.App{}, flag.NewFlagSet("test", 0), nil).Context,
		BeaconCaller,
		selfAddress,
		selfCert,
		gatewayAddress,
		nil, /*optional mux*/
		allowedOrigins,
		enableDebugRPCEndpoints,
		ctx.Uint64("grpc-max-msg-size"),
	)
```

File: shared/gateway/gateway_test.go (hunk 7 above)
Comment (rkapka) [reply shared/gateway/gateway_test.go:48]: Let's keep it the way it is then. It will impact other services, so if we want to add logging it should be in another PR.

File: shared/gateway/gateway_test.go (hunk 12 above)
Comment (rkapka) [reply shared/gateway/gateway_test.go:21]: The usual way create a context in tests is through `ctx := context.Background()`.

File: shared/gateway/gateway.go (hunk 2 above)
Comment (rkapka) [reply shared/gateway/gateway.go]: Can you please follow the same convention and use `iota`?

File: shared/gateway/gateway_test.go (hunk 12 above)
Comment (ahadda5) [reply shared/gateway/gateway_test.go:21]: The thing is the cli library allows for the fetching of values from the flags efficiently. 
e.g. 
`gatewayPort := ctx.Int(flags.GRPCGatewayPort.Name)`

will be something like  this ; done directly which albeit being intuitive was not the way currently used in the code. 
`gatwayPort :=  flags.GRPCGatewayPort.Value `

File: shared/gateway/gateway.go (hunk 2 above)
Comment (ahadda5) [reply shared/gateway/gateway.go]: Done

**Review by rkapka — APPROVED** [review]
